xpcPlugin/MessageHandlers.cpp

```diff
@@ -933,15 +935,65 @@ namespace XPC
  			std::string comm = std::string((char*)buffer + pos, len);
  			pos += len;
 
-  			DataManager::Execute(comm);
- 			Log::FormatLine(LOG_DEBUG, "COMM", "Execute command %s", comm.c_str());
+  			DataManager::MomentaryCommand(comm);
+ 			Log::FormatLine(LOG_DEBUG, "COMM", "Execute momentary command %s", comm.c_str());
  		}
  		if (pos != size)
  		{
  			Log::WriteLine(LOG_ERROR, "COMM", "ERROR: Command did not terminate at the expected position.");
  		}
  	}
 
+	void MessageHandlers::HandleBCom(const Message& msg)
+	{
+		Log::FormatLine(LOG_TRACE, "BCOM", "Request to execute BCOM command received (Conn %i)", connection.id);
+		const unsigned char* buffer = msg.GetBuffer();
+		std::size_t size = msg.GetSize();
+		std::size_t pos = 5;
+		while (pos < size)
+		{
+			unsigned char len = buffer[pos++];
+			if (pos + len > size)
+			{
+				break;
+			}
+			std::string comm = std::string((char*)buffer + pos, len);
+			pos += len;
+
+			DataManager::BeginCommand(comm);
+			Log::FormatLine(LOG_DEBUG, "BCOM", "Begin command %s", comm.c_str());
+		}
+		if (pos != size)
+		{
+			Log::WriteLine(LOG_ERROR, "BCOM", "ERROR: Command did not terminate at the expected position.");
+		}
+	}
+
+	void MessageHandlers::HandleECom(const Message& msg)
+	{
+		Log::FormatLine(LOG_TRACE, "ECOM", "Request to execute ECOM command received (Conn %i)", connection.id);
+		const unsigned char* buffer = msg.GetBuffer();
+		std::size_t size = msg.GetSize();
+		std::size_t pos = 5;
+		while (pos < size)
+		{
+			unsigned char len = buffer[pos++];
+			if (pos + len > size)
+			{
+				break;
```

Comment:
I don't understand this condition. Doesn't it cause us to skip the final command?

Reply:
The network data structure contains the string length, immediately followed by a variable-length string with the command name. This condition checks to make sure the string length doesn't exceed the bounds of the incoming packet. (e.g. if someone sent in a packet that's 10 bytes long, but incorrectly specified "20" in the string length field, you'd index past the end of the buffer if you didn't have this check).